### ISSUE TITLE
PREWHERE filters may be ignored when row policy used

### ISSUE THREAD

**Issue by MyroTk** [ClickHouse/ClickHouse#40956]
**Describe the unexpected behaviour**
Using a WITH statement when row policy is applied to the table leads to unexpected result where rows are not getting filtered how they should be.

**How to reproduce**
* Which ClickHouse server version to use
`ClickHouse server version 22.8.2 revision 54460`

* Queries to run that lead to unexpected result
```
): CREATE TABLE table0
(
    `x` Int16,
    `y` String
)
ENGINE = MergeTree
ORDER BY x
SETTINGS index_granularity = 8192

): INSERT INTO table0(y, x) VALUES ('A',1), ('B',2), ('C',3)

): CREATE ROW POLICY pol0 ON default.table0 FOR SELECT USING x >= 0 TO default

): WITH Test AS
    (
        SELECT
            multiIf(y = 'Aaaaaaaaaaaa', 'Bbbbbbbbbbbb', y) AS y,
            x
        FROM table0
        WHERE y IN ('A', 'C')
    )
SELECT DISTINCT (y, x)
FROM Test

Query id: 0b494596-392a-47cf-8587-239561219844

┌─tuple(y, x)─┐
│ ('A',1)     │
│ ('B',2)     │
│ ('C',3)     │
└─────────────┘
```

**Expected behavior**
Only 'A' and 'C' entries should appear, 'B' should be excluded by the `WHERE y IN ('A', 'C')` clause.

**Comment by MyroTk** [comment]
This only occurs under the following conditions:
- multiIf column is named the same as an existing column.
- The row policy is not filtering any rows.
- `SELECT DISTINCT` is selecting multiple columns. `SELECT DISTINCT tuple(y,)` works as expected.
- The data must have been inserted in the same block. If the rows are inserted separately, the query returns the expected result.
- The length of strings in the multiIf makes a difference.  It is not clear when this will cause the issue because if the strings are too short or too long, the query returns the correct result.  For example, `multiIf(y = 'Aa', 'Bb', y) AS y,` and `multiIf(y = 'Aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa', 'Bbbbbbbbbbbb', y) AS y,` follows the expected behavior.

**Comment by filimonov** [comment]
It related to prewhere (workaround is optimize_move_to_prewhere=0) , and not related to with. 

Simpler repro:
```
select * FROM table0 PREWHERE if(y = 'Aaaaaaaaaaa', 'Bbbbbbbbbbbbb', y) IN ('foo');
```

**Comment by azat** [comment]
>Maybe related https://github.com/ClickHouse/ClickHouse/pull/37165

Yeah, kind of, but not directly it just unhide the bug that was there before, anyway see the referenced PR.

**Comment by filimonov** [comment]
Fixed in https://github.com/ClickHouse/ClickHouse/pull/42126

Released in 22.12, backported to 22.8.12
